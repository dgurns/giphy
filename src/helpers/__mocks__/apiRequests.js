export const searchGifs = jest.fn(() =>
  Promise.resolve([
    {
      images: {
        fixed_height_downsampled: {
          url: 'downsampledUrl'
        },
        original: {
          url: 'originalUrl'
        }
      },
      title: 'First gif',
      id: 'abc123'
    },
    {
      images: {
        fixed_height_downsampled: {
          url: 'downsampledUrl'
        },
        original: {
          url: 'originalUrl'
        }
      },
      title: 'Second gif',
      id: '321cba'
    }
  ])
);
