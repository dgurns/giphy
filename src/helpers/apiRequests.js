import axios from 'axios';

const GIPHY_API_ENDPOINT = 'https://api.giphy.com/v1/gifs/search';
const GIPHY_API_KEY = 'bKenxoSj2MMDE6GXV1icMmPfJZZyDWnM';
const NUMBER_OF_RESULTS = 10;

export const searchGifs = async (query, resultsPage = 1) => {
  try {
    const response = await axios.get(GIPHY_API_ENDPOINT, {
      params: {
        api_key: GIPHY_API_KEY,
        q: query,
        limit: NUMBER_OF_RESULTS,
        offset: NUMBER_OF_RESULTS * (resultsPage - 1)
      }
    });
    return response.data.data;
  } catch (e) {
    throw new Error('Error fetching gifs. Please try again.');
  }
};
