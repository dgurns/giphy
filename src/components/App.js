import React, { Component } from 'react';
import { searchGifs } from '../helpers/apiRequests';
import Pagination from './Pagination';
import Modal from './Modal';

class App extends Component {
  state = {
    searchQuery: '',
    results: [],
    resultsPage: 1,
    error: '',
    modalVisible: false,
    selectedGif: null
  };

  componentDidUpdate(prevProps, prevState) {
    if (prevState.resultsPage !== this.state.resultsPage) {
      this.fetchGifs();
    }
  }

  onPaginationClicked = (increment = true) => {
    const { resultsPage } = this.state;
    const updatedResultsPage = increment ? resultsPage + 1 : resultsPage - 1;
    this.setState({ resultsPage: updatedResultsPage });
  };

  onSearchClicked = () => {
    this.setState({ resultsPage: 1 }, () => this.fetchGifs());
  };

  async fetchGifs() {
    const { searchQuery, resultsPage } = this.state;
    this.setState({ error: '', results: [] });
    try {
      const results = await searchGifs(searchQuery, resultsPage);
      this.setState({ results });
    } catch (e) {
      this.setState({ error: e.message });
    }
  }

  onGifClicked = gif => {
    this.setState({ selectedGif: gif, modalVisible: true });
  };

  renderModal() {
    const { modalVisible, selectedGif } = this.state;
    if (modalVisible && selectedGif) {
      const { images, title } = selectedGif;
      return (
        <Modal
          onClose={() =>
            this.setState({ modalVisible: false, selectedGif: null })
          }
        >
          <img className="modal__image" src={images.original.url} alt={title} />
        </Modal>
      );
    }
  }

  render() {
    const { searchQuery, results, resultsPage, error } = this.state;

    return (
      <div className="app">
        <h1>Giphy Search</h1>
        <div className="app__search">
          <input
            type="text"
            placeholder="Enter search term..."
            value={searchQuery}
            onChange={e => this.setState({ searchQuery: e.target.value })}
          />
          <button onClick={this.onSearchClicked.bind(this)}>Search</button>
        </div>
        <div className="app__results-container">
          {results.map(gif => (
            <img
              className="app__result-item"
              src={gif.images.fixed_height_downsampled.url}
              alt={gif.title}
              onClick={() => this.onGifClicked(gif)}
            />
          ))}
          {error}
        </div>
        {results.length > 0 && (
          <Pagination
            currentPage={resultsPage}
            onPageForward={() => this.onPaginationClicked()}
            onPageBack={() => this.onPaginationClicked(false)}
          />
        )}
        {this.renderModal()}
      </div>
    );
  }
}

export default App;
