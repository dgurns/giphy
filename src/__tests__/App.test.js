import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import App from '../components/App';
import { searchGifs } from '../helpers/apiRequests';

jest.mock('../helpers/apiRequests');

describe('<App /> component', () => {
  beforeEach(() => {
    searchGifs.mockClear();
  });

  it('should render without crashing', () => {
    const { container } = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call searchGifs API request with given term when search is clicked', () => {
    const { getByText, getByPlaceholderText } = render(<App />);
    const searchField = getByPlaceholderText(/Enter search term/i);
    fireEvent.change(searchField, { target: { value: 'dog' } });
    fireEvent.click(getByText('Search'));
    expect(searchGifs).toHaveBeenCalledTimes(1);
    expect(searchGifs).toHaveBeenNthCalledWith(1, 'dog', 1);
  });

  it('should display the modal when a gif is clicked', async () => {
    const { getByText, getByAltText, getByPlaceholderText } = render(<App />);
    const searchField = getByPlaceholderText(/Enter search term/i);
    fireEvent.change(searchField, { target: { value: 'dog' } });
    fireEvent.click(getByText('Search'));
    await wait(() => {
      fireEvent.click(getByAltText('First gif'));
    });
    const modalCloseButton = getByText('Close');
    expect(modalCloseButton).toBeTruthy();
  });

  it('should request the next page of API results when the "Next" button is clicked', async () => {
    const { getByText, getByPlaceholderText } = render(<App />);
    const searchField = getByPlaceholderText(/Enter search term/i);
    fireEvent.change(searchField, { target: { value: 'dog' } });
    fireEvent.click(getByText('Search'));
    expect(searchGifs).toHaveBeenCalledTimes(1);
    expect(searchGifs).toHaveBeenNthCalledWith(1, 'dog', 1);
    await wait(() => {
      fireEvent.click(getByText('>'));
    });
    expect(searchGifs).toHaveBeenCalledTimes(2);
    expect(searchGifs).toHaveBeenNthCalledWith(2, 'dog', 2);
    fireEvent.click(getByText('>'));
    expect(searchGifs).toHaveBeenCalledTimes(3);
    expect(searchGifs).toHaveBeenNthCalledWith(3, 'dog', 3);
  });
});
