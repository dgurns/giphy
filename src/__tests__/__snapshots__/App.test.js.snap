// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> component should render without crashing 1`] = `
<div
  class="app"
>
  <h1>
    Giphy Search
  </h1>
  <div
    class="app__search"
  >
    <input
      placeholder="Enter search term..."
      type="text"
      value=""
    />
    <button>
      Search
    </button>
  </div>
  <div
    class="app__results-container"
  >
    
  </div>
</div>
`;
